src/inmanta/ast/constraint/expression.py

```diff
@@ -514,29 +514,57 @@ class ArithmeticOperator(BinaryOperator):
     __slots__ = ()
 
     def _bin_op(self, arg1: object, arg2: object) -> object:
-        if not isinstance(arg1, (int, float)):
-            raise TypingException(self, f"The {self.get_name()} operator in not supported for {arg1} of type {type(arg1)}.")
-        if not isinstance(arg2, (int, float)):
-            raise TypingException(self, f"The {self.get_name()} operator in not supported for {arg2} of type {type(arg2)}.")
-        return self._arithmetic_op(arg1, arg2)
+        result: object = self._execute_operator(arg1, arg2)
+        if result is NotImplemented:
+            raise TypingException(
+                self,
+                (
+                    f"Unsupported operand type(s) for {self.get_name()}:"
+                    f" '{type(arg1).__name__}' ({repr(arg1)}) and '{type(arg2).__name__}' ({repr(arg2)})"
+                ),
+            )
+        return result
+
+    def _execute_operator(self, arg1: object, arg2: object) -> object:
+        """
+        Validate and execute this operator given two operands. When validation fails, may raise its
+        own custom TypingException or return the special value NotImplemented, which will result in a generic TypingException.
+
+        This class adds an implementation for numbers. This may be extended by operators that support additional types.
+        """
+        if isinstance(arg1, (int, float)) and isinstance(arg2, (int, float)):
+            return self._arithmetic_op(arg1, arg2)
+        return NotImplemented
 
     @abstractmethod
     def _arithmetic_op(self, arg1: int | float, arg2: int | float) -> int | float:
         """
-        The implementation for this ArithmeticOperator, excluding the type validation.
-        Type validation is done in the _bin_op() method.
+        The implementation for this ArithmeticOperator when applied to numbers, excluding the type validation.
+        Type validation is done in the _execute_operator() method.
+        Concrete implementations may widen the type signature.
         """
         raise NotImplementedError()
 
 
+T = TypeVar("T", int | float, str)
+
+
 class Plus(ArithmeticOperator):
     __slots__ = ()
     __op = "+"
 
     def __init__(self, op1: ExpressionStatement, op2: ExpressionStatement) -> None:
         ArithmeticOperator.__init__(self, "plus", op1, op2)
 
-    def _arithmetic_op(self, arg1: int | float, arg2: int | float) -> int | float:
+    def _execute_operator(self, arg1: object, arg2: object) -> object:
+        if isinstance(arg1, str):
+            if isinstance(arg2, str):
+                return self._arithmetic_op(arg1, arg2)
+            # raise more representative error to override generic one from ArithmeticOperator
+            raise TypingException(self, f"Can only concatenate str (not '{type(arg2).__name__}' ({repr(arg2)})) to str")
```

Comment:
We will only get this more specific exception message if arg1 is a string and arg2 isn't. But not the other way around. Is that intentional?

Reply:
Sort of, but I'm open to suggestions. This happens to be how Python reports it and I followed that:
```ipython
>>> "hello" + 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only concatenate str (not "int") to str
>>> 1 + "hello"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'int' and 'str'
>>>
```

I feel like it makes sense from the point of view of `x + y` meaning "add `y` to `x`". The type of `y` would then determine the type of the operation, and the type of `x` would determine whether the operation can be applied.

Reply:
I find this a bit strange. On the other hand, both error messages are rather clear with respect to what is going wrong. So it's not really that important. You can ignore this comment.

Reply:
Ok. We can always refine if we get a better idea.